indexing-service/src/main/java/org/apache/druid/indexing/common/actions/LocalTaskActionClient.java

```diff
@@ -72,14 +77,35 @@ public <RetType> RetType submit(TaskAction<RetType> taskAction)
 
     final long performStartTime = System.currentTimeMillis();
     final RetType result = taskAction.perform(task, toolbox);
-    emitTimerMetric("task/action/run/time", System.currentTimeMillis() - performStartTime);
+    emitTimerMetric("task/action/run/time", taskAction, System.currentTimeMillis() - performStartTime);
     return result;
   }
 
-  private void emitTimerMetric(final String metric, final long time)
+  private void emitTimerMetric(final String metric, final TaskAction<?> action, final long time)
   {
     final ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
     IndexTaskUtils.setTaskDimensions(metricBuilder, task);
+    final String actionType = getActionType(toolbox.getJsonMapper(), action);
+    if (actionType != null) {
+      metricBuilder.setDimension("taskActionType", actionType);
+    }
     toolbox.getEmitter().emit(metricBuilder.build(metric, Math.max(0, time)));
   }
+
+  @Nullable
+  static String getActionType(final ObjectMapper jsonMapper, final TaskAction<?> action)
+  {
+    try {
+      final Map<String, Object> m = jsonMapper.convertValue(action, JacksonUtils.TYPE_REFERENCE_MAP_STRING_OBJECT);
```

Comment:
This seems a bit overkill and would add an extra serde step on completion of every task action.
Should we expose a `getType()` method on the TaskAction instead or maybe emit the class name as the type (similar to coordinator duty dimension in `coordinator/run/time`)?

Reply:
I thought about that, but felt that adding `getType()` to all TaskActions would be cumbersome since there are many kinds of actions, and that the perf impact of doing this `convertValue` is low, since actions tend to be pretty heavy weight compared to `convertValue`.

Reply:
Yeah, actions themselves are pretty heavy. Thanks for the clarification.